Abins: improve handling of isotopic cross-sections

ajjackson opened the pull request:
### Description of work
#### Summary of work

- [x] introduce unit test that error should be raised for Zn65
- [x] modify code to raise ValueError in such cases
- [x] introduce unit test that Zn_30_65.39 should be identified as isotopic mixture
- [x] modify code to detect when standard mass is used


#### Purpose of work

__Primary issue__:
Unstable isotopes such as Zn65 currently give a NaN cross-section. We should ensure that unsupported isotopes give a noisy error instead.

__Secondary issue__:
We are seeing Zn65 because data with mass 65.39 is being inappropriately rounded to this exotic isotope, rather than identified as being the standard isotopic mixture. Such cases should be detected more robustly.

Fixes #37735 


**Report to:** Jeff Armstrong


#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->

### To test:
The changes are covered by unit tests and system tests

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

ajjackson committed:
Create failing test: Abins get_cross_section for Zn65

This isotope currently gives a NaN cross-section. We should ensure
that unsupported isotopes give a noisy error instead.

ajjackson committed:
Implement ValueError for NaN cross-section data

ajjackson committed:
Fix missing return value, test good cases

ajjackson commented:
Currently Abins identifies if the mass is a standard isotopic mixture by comparing against `Atom(symbol=symbol).mass` with a threshold of 0.01.

In the case of zinc, the standard mass has been revised by a difference of more than 0.01, so CASTEP writes mass values that are not detected by Abins as being the standard.

A more robust workflow could look at which is _nearest_ between the standard mass and another known isotope.

I have found a quirk the `Atom` API though:

```
In [44]: Atom(symbol="Zn").mass
Out[44]: 65.409

In [45]: Atom(symbol="Zn", a_number=82).mass
Out[45]: 81.95484

In [46]: Atom(symbol="Zn", a_number=83).mass
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[46], line 1
----> 1 Atom(symbol="Zn", a_number=83).mass

RuntimeError: Failed to find an atom with symbol=Zn and a=83

In [47]: Atom(symbol="Zn", a_number=20).mass
Out[47]: 53.99295
```

Mass values that are too high raise an error, but masses that are too _low_ do not!

ajjackson committed:
Begin refactoring Abins isotope handling

- remove "substitution" logic branch
- instead delegate isotope detection and workspace naming to a new class
- in this commit a new Atom is still (unnecessarily) spawned to get cross-section

ajjackson committed:
Continue refactor: streamline _create_workspace

We can now pass on the AtomInfo object rather than its consitituent data.

ajjackson committed:
Continue refactor: move AtomInfo into _fill_s_workspace methods

ajjackson committed:
Simplify get_cross_section from an established AtomInfo

ajjackson committed:
Handle neutron data for atoms where one isotope = mixture

In cases such as F, one isotope is so dominant that essentially we
would always use the properties of the standard mixture. Oddly, in
this case Mantid will not find neutron data if the atomic mass was
provided.

i.e. Atom('F').neutron() works, but Atom('F', 19).neutron() returns a
dict full of NaN, despite them having exactly the same mass.

ajjackson committed:
Update Abins2D to use AtomInfo class

ajjackson committed:
Handle some more Abins isotope challenging cases

Abins input files usually give the element symbol and mass. These
masses often do not match exactly to those in Mantid's data.

- In the case that the nearest isotope has no data, but the "mixture"
  is _quite_ near the target mass (0.01 amu, for now), log a warning and
  use the mixture.

  This helps handle nasty cases like 127I where Mantid has _slightly_
  different masses for the pure isotope and mixture, but essentially
  it is a 100% mixture and there is no neutron-scattering data for the
  isolated isotope.

- If the nearest isotope has no data and the standard mixture is too
  far away, raise an error.

ajjackson commented:
For an example of a really nasty case that requires some logic to get through:

- Mantid's `Atom.cpp` defines a standard I53 with mass 126.904470 and, among many others, an isotope 127I53 with abundance 100% and mass 126.904468
- The mass assigned to "standard" iodine by CASTEP is 126.903999; presumably these codebases are using reference data from different years.

To resolve this, we
- check if the requested mass is closer to the Mantid reference mass of 127I53 or I53;
- in this case the 127I53 mass is actually closer so we assume this could be what they wanted.
- Then we check if data is available; the coherent scattering length is NaN so it is not.
- Because we are within 0.01 of the standard mass, we assume this was the original intent, raise a warning, and use the I53 data.

ajjackson committed:
Direct unit test for AtomInfo

ajjackson committed:
Test refactor: move error check closer to source

The error is now raised by the data class when we try to access a
property that uses Atom.

As such it is not really a get_cross_section test any more and can be
moved out.

ajjackson committed:
Add release note

ajjackson commented:
The docs failure might be connected to 
> 17:04:07 release/v6.12.0/indirect_geometry:3: WARNING: Bullet list ends without a blank line; unexpected unindent. [docutils]

but I'm not sure what to do about that. The pre-commit hooks will not let me append another newline to the release note!

ajjackson committed:
Tweak release note format

ajjackson commented:
Thanks @robertapplin for figuring out the docs formatting 🎉

adriazalvarez commented on scripts/abins/abinsalgorithm.py:
```diff
@@ -24,10 +27,66 @@
+import abins.parameters
+
+
+@dataclasses.dataclass
```
Does this dataclass belong in this file? What do you think of moving it to a separate file?

ajjackson replied:
No strong opinion. It is used in two files, but abins2 is always going to import from abinsalgorithm so there is no danger of circular references. (If anything the danger increases by introducing a new module.)

I'm happy to move it if you think it will be tidier.

adriazalvarez replied:
Design wise is fine, I was more thinking that in abins library every filename sort of contains a class with the same name as the file. If the atoms info is an independent entity, probably it would be stylistically more clean to have it in a separate file.

adriazalvarez commented on scripts/abins/abinsalgorithm.py:
```diff
@@ -24,10 +27,66 @@
+        nearest_isotope = Atom(symbol=self.symbol, a_number=nearest_int)
+        standard_mix = Atom(symbol=self.symbol)
+
+        if abs(nearest_isotope.mass - standard_mix.mass) < 1e-12:
```
Could we move the epsilon to be a constant? I know is very unlikely this number has to be changed, but to make the code look more homogeneous.

adriazalvarez commented on scripts/abins/abinsalgorithm.py:
```diff
@@ -24,10 +27,66 @@
+            if abs(self.mass - standard_mix.mass) > MASS_EPS:
+                raise ValueError(f"Could not find suitable isotope data for {self.symbol} with mass {self.mass}.")
+
+            logger.warning(f"Standard mass {standard_mix.mass} is close enough, will " f"use data for this isotope mixture.")
```
```suggestion
            logger.warning(f"Standard mass {standard_mix.mass} is close enough, will use data for this isotope mixture.")
```
I think the precommit is modifying this string to be in a single line.

ajjackson replied:
140 characters is so wiiiiiiide

adriazalvarez reviewed (changes requested):
The changes are sensible and adding the atominfo dataclass handles better edge cases and also simplifies the code in some parts. Additionally, test cases are added. 
I have very small comments but otherwise looks fine to me.

ajjackson committed:
Update scripts/abins/abinsalgorithm.py

Co-authored-by: Adri Diaz <146007827+adriazalvarez@users.noreply.github.com>

ajjackson committed:
Move AtomInfo to separate file

ajjackson committed:
Tighten epsilon for "same" mass values, re-use in atominfo

This gets rid of a hard-coded epsilon. Strictly these two comparisons
are not quite doing the same thing, but this precision should be good
for both.

adriazalvarez reviewed (approved):
Thanks for addressing my comments.
